File: addons/terrabrush/Scripts/Tools/WaterFlowTool.cs
#if TOOLS
using Godot;

namespace TerraBrush;

public class WaterFlowTool : ToolBase {
    private Vector2 _previousWaterMousePosition = Vector2.Zero;
    private Vector2 _previousWaterMouseDirection = Vector2.Zero;

    public WaterFlowTool(TerraBrush terraBrush) : base(terraBrush) {}

    protected override ImageTexture GetToolCurrentImageTexture(ZoneResource zone) {
        return zone.WaterTexture;
    }

    public override void Paint(TerrainToolType toolType, Image brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
        if (_terraBrush.WaterDefinition == null) {
            return;
        }

        ForEachBrushPixel(brushImage, brushSize, imagePosition, (imageZoneInfo, pixelBrushStrength) => {
            var currentPixel = imageZoneInfo.Image.GetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y);

            if (currentPixel.R > 0.0) {
                var direction = _previousWaterMousePosition.DirectionTo(imagePosition) * -1;

                if (direction == Vector2.Zero) {
                    direction = _previousWaterMouseDirection;
                } else {
                    direction = new Vector2(
                        Mathf.Lerp(_previousWaterMouseDirection.X, direction.X, 0.1f),
                        Mathf.Lerp(_previousWaterMouseDirection.Y, direction.Y, 0.1f)
                    );
                    _previousWaterMouseDirection = direction;
                }

                var newColor = toolType == TerrainToolType.WaterFlowAdd ? new Color(currentPixel.R, Mathf.Remap(direction.X, -1, 1, 0, 1), Mathf.Remap(direction.Y, -1, 1, 0, 1), 1) : new Color(currentPixel.R, 0.5f, 0.5f, 1);
                var newValue = new Color(
                    Mathf.Lerp(currentPixel.R, newColor.R, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.G, newColor.G, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.B, newColor.B, pixelBrushStrength * brushStrength),
                    Mathf.Lerp(currentPixel.A, newColor.A, pixelBrushStrength * brushStrength)
                );
                imageZoneInfo.Image.SetPixel(imageZoneInfo.ZoneInfo.ImagePosition.X, imageZoneInfo.ZoneInfo.ImagePosition.Y, newValue);
            }
        });

        _terraBrush.TerrainZones.UpdateWaterTextures();
        _previousWaterMousePosition = imagePosition;
    }
}
#endif

File: src/editor_tools/water_flow_tool.cpp
#include "water_flow_tool.h"

using namespace godot;

void WaterFlowTool::_bind_methods() {}

WaterFlowTool::WaterFlowTool() {}

WaterFlowTool::~WaterFlowTool() {}

bool WaterFlowTool::getApplyResolution() {
    return true;
};

Ref<ImageTexture> WaterFlowTool::getToolCurrentImageTexture(Ref<ZoneResource> zone) {
    return zone->get_waterTexture();
};

void WaterFlowTool::paint(TerrainToolType toolType, Ref<Image> brushImage, int brushSize, float brushStrength, Vector2 imagePosition) {
    if (_terraBrush->get_waterDefinition().is_null()) {
        return;
    }

    forEachBrushPixel(brushImage, brushSize, imagePosition, ([&](ImageZoneInfo &imageZoneInfo, float pixelBrushStrength) {
        Color currentPixel = imageZoneInfo.image->get_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y);

        if (currentPixel.r > 0.0) {
            Vector2 direction = _previousWaterMousePosition.direction_to(imagePosition) * -1;

            if (direction == Vector2(0, 0)) {
                direction = _previousWaterMouseDirection;
            } else {
                direction = Vector2(
                    Math::lerp(_previousWaterMouseDirection.x, direction.x, 0.1f),
                    Math::lerp(_previousWaterMouseDirection.y, direction.y, 0.1f)
                );
                _previousWaterMouseDirection = direction;
            }

            Color newColor = toolType == TerrainToolType::TERRAINTOOLTYPE_WATERFLOWADD ? Color(currentPixel.r, Math::remap(direction.x, -1, 1, 0, 1), Math::remap(direction.y, -1, 1, 0, 1), 1) : Color(currentPixel.r, 0.5f, 0.5f, 1);
            Color newValue = Color(
                Math::lerp(currentPixel.r, newColor.r, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.g, newColor.g, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.b, newColor.b, pixelBrushStrength * brushStrength),
                Math::lerp(currentPixel.a, newColor.a, pixelBrushStrength * brushStrength)
            );
            imageZoneInfo.image->set_pixel(imageZoneInfo.zoneInfo.imagePosition.x, imageZoneInfo.zoneInfo.imagePosition.y, newValue);
        }
    }));

    _terraBrush->get_terrainZones()->updateWaterTextures();
    _previousWaterMousePosition = imagePosition;
};